pkg/api/http_server.go

```diff
@@ -406,45 +413,56 @@ func (hs *HTTPServer) Run(ctx context.Context) error {
 	hs.log.Info("HTTP Server Listen", "address", listener.Addr().String(), "protocol",
 		hs.Cfg.Protocol, "subUrl", hs.Cfg.AppSubURL, "socket", hs.Cfg.SocketPath)
 
-	var wg sync.WaitGroup
-	wg.Add(1)
-
-	// handle http shutdown on server context done
-	go func() {
-		defer wg.Done()
-
-		<-ctx.Done()
-		if err := hs.httpSrv.Shutdown(context.Background()); err != nil {
-			hs.log.Error("Failed to shutdown server", "error", err)
-		}
-	}()
-
 	switch hs.Cfg.Protocol {
 	case setting.HTTPScheme, setting.SocketScheme:
-		if err := hs.httpSrv.Serve(listener); err != nil {
-			if errors.Is(err, http.ErrServerClosed) {
-				hs.log.Debug("server was shutdown gracefully")
-				return nil
+		go func() {
+			if err := hs.httpSrv.Serve(listener); err != nil {
+				if errors.Is(err, http.ErrServerClosed) {
+					hs.log.Debug("server was shutdown gracefully")
+					close(hs.errs)
+					return
+				}
+				hs.errs <- err
 			}
-			return err
-		}
+		}()
 	case setting.HTTP2Scheme, setting.HTTPSScheme:
-		if err := hs.httpSrv.ServeTLS(listener, hs.Cfg.CertFile, hs.Cfg.KeyFile); err != nil {
-			if errors.Is(err, http.ErrServerClosed) {
-				hs.log.Debug("server was shutdown gracefully")
-				return nil
+		go func() {
+			if err := hs.httpSrv.ServeTLS(listener, hs.Cfg.CertFile, hs.Cfg.KeyFile); err != nil {
+				if errors.Is(err, http.ErrServerClosed) {
+					hs.log.Debug("server was shutdown gracefully")
+					close(hs.errs)
+					return
+				}
+				hs.errs <- err
 			}
-			return err
-		}
+		}()
 	default:
 		panic(fmt.Sprintf("Unhandled protocol %q", hs.Cfg.Protocol))
 	}
 
-	wg.Wait()
+	return nil
+}
+
+func (hs *HTTPServer) running(ctx context.Context) error {
+	select {
+	case err, ok := <-hs.errs:
+		if !ok {
+			return nil
+		}
+		return err
+	case <-ctx.Done():
+	}
 
 	return nil
 }
 
+func (hs *HTTPServer) stop(failureReason error) error {
+	if err := hs.httpSrv.Shutdown(context.Background()); err != nil {
+		return fmt.Errorf("failed to shutdown server: %w", err)
+	}
+	return failureReason
+}
+
```

Comment:
@mildwonkey this is what i was thinking it might look like based on the loki example:
```suggestion
func (hs *HTTPServer) stop(_ error) error {
	if err := hs.httpSrv.Shutdown(context.Background()); err != nil {
		return fmt.Errorf("failed to shutdown server: %w", err)
	}
	return nil
}

```